tensorflow/compiler/jit/xla_gpu_device.cc

```diff
@@ -46,7 +47,7 @@ Status XlaGpuDeviceFactory::CreateDevices(const SessionOptions& options,
 
   std::unique_ptr<XlaDevice> device;
   Status status =
-      XlaDevice::Create("CUDA", DEVICE_XLA_GPU, 0, DEVICE_GPU_XLA_JIT, options,
+      XlaDevice::Create(GPU_PLATFORM_NAME, DEVICE_XLA_GPU, 0, DEVICE_GPU_XLA_JIT, options,
```

Comment:
@hawkinsp Again I don't think a new class of device for AMD GPU is a viable choice, due to how XLA is coupled with GPU common runtime. And how GPU-supported operators are currently registered. 

In the implementation of `xla_op_registry` / `xla_launch_op` / `mark_for_compilation_pass`, `DEVICE_GPU_XLA_JIT` & `DEVICE_XLA_GPU` are tightly couple with `DEVICE_GPU` in GPU common runtime. And operator implementations in `tensorflow/core/kernels` or `contrib` also register themselves be executed on `DEVICE_GPU`.

Shall a new class of device for AMD GPU is introduced, all these places would have to be altered. We would have duplicated codes to simply register ops for the new device class. And due to the fact user codes could physically specify devices a op be executed, adding new device classes would surely mean such user codes have to be modified which is not feasible.

Finally, there are other places in Python APIs where `device_type` of a device be checked. Adding a new device class could mean those places have to be amended as well. And again that defeats the purpose of letting user codes stay as is, and be able to run on implementations from different vendors.

Reply:
I think it's important to distinguish between "GPU" and "XLA_GPU". They are to some extent completely separate issues because they share almost nothing (although both can call XLA). Let's discuss the "GPU" case in the other comment and resolve that first.

For the XLA_GPU case, I am confident that separating the two is feasible. The coupling between XLA_GPU_JIT and GPU is limited to a small number of registrations that say how they connect. It seems quite feasible to add another backend. As I observed above, this device is really only used by tests anyway, so compatibility isn't necessary.

Another concern here is that GOOGLE_CUDA is not necessarily defined when using XLA_GPU — XLA does not require a CUDA compiler to generate GPU code. The opensource configure script conflates "targeting an NVidia GPU" with "having a CUDA compiler", but they are not actually linked when using XLA and we depend on this behavior for some of our internal XLA testing. So I prefer that we separate XLA_GPU from XLA_AMDGPU, irrespective of what we choose to do for the TF core GPU device.

Reply:
FWIW I agree that it would be great for users to be able to have AMD and nvidia GPU support in a single binary, and in contrast pretty bad if we decided now that forevermore you'd never be able to build a single binary with support for both.

An example of where not being able to "support both" has bitten us in the past is with CPU optimizations.  C++ makes it very hard to compile one file for two ISA extensions (say, SSE2 and AVX), so by and large we don't do it.  This means that anyone who downloads the tensorflow package gets...something, but not what's optimal.

One could say, well, we should have tensorflow-sse, tensorflow-sse2, tensorflow-sse3, tensorflow-ssse3, tensorflow-avx, tensorflow-avx2, tensorflow-avx512, ...  But it should be clear how quickly that gets out of hand.  Plus we're adding another dimension with the gpu bit.

> Shall a new class of device for AMD GPU is introduced, all these places would have to be altered. We would have duplicated codes to simply register ops for the new device class. 

They're macros anyway, this doesn't seem like a big deal to change?

> And due to the fact user codes could physically specify devices a op be executed, adding new device classes would surely mean such user codes have to be modified which is not feasible.

This is the bigger issue from my perspective.  But it's not clear to me that it's impossible to come up with a way to allow user code to be unaffected while still accomplishing what hawkinsp is trying to do.

Reply:
@hawkinsp I agree we can separate the topic of "GPU" and "XLA_GPU", and I agree for "XLA_GPU" the separation could be possible and imposes less impact to TF core "GPU" device. Let's resolve "GPU" case first.

Reply:
@jlebar / @hawkinsp Thus far I think whether "having one binary to support both CUDA and ROCm" is a true merit and worthwhile become a requirement remains unknown. In #20277 I was asked to change `configure.py` so CUDA / SYCL / ROCm are mutually exclusive at configure time.

Would it be possible to consider keeping the current form of having CUDA / ROCm be mututally exclusive at configure/compile-time, and having ROCm-based TensorFlow be shipped in a different PyPI package, say `tensorflow-rocm`? Actually @parallelo and I are already working on the infrastructure to publish such package to PyPI.

We can leave the discussion of "one binary to support both CUDA and ROCm" for future versions when "hip-clang" becomes mature enough. By then we can discuss about about to merge CROSSTOOL for both platforms, how to merge kernel sections and emissions of hidden runtime function to register GPU kernels.

Reply:
I agree #20277 did point you towards enforcing that the two options are not simultaneously set, but that was mostly because it didn't make sense to do so (the assertion was that the two did not work at the same time). I also suspect the deployment question is separate: even if you *can* build a binary that supports multiple GPUs, you don't necessarily have to deploy that binary in that form via PyPI.

I'm fine if the two are forced to be mutually exclusive at configure time for now.

I'd just like to agree on how the TF devices are named and how the code is structured now before we check in much more, because changing it later will be very hard.

Reply:
@hawkinsp Thanks for all the comments. Based on the discussion I'll revise the PR so:

- in XLA, a new class of device for AMD GPU be introduced: `XLA_AMDGPU`
- in XLA, no conditioning based on `GOOGLE_CUDA` would be made

This allows XLA to possibly support emitting GPU kernels for both NVPTX and AMDGPU targets.


- in GPU common runtime, keep "GPU" be either CUDA or ROCm via the mutually exclusive configure-time option

Changes to `gpu_init` would likely be removed from this PR and be filed as a separate one. In fact I've also got some other PRs to be filed to decouple CUDA (ex: `CudaGpuId`) from GPU common runtime.

Reply:
This last proposal sounds good to me.

(I'll also rename XLA_GPU to XLA_NVGPU to minimize confusion, too, in a subsequent change.)